great_expectations/data_context/types/base.py

```diff
@@ -148,14 +180,109 @@ class Meta:
 
     assets = fields.Dict(
         keys=fields.Str(),
-        values=fields.Nested(AssetConfigSchema),
+        values=fields.Nested(AssetConfigSchema, required=False, allow_none=True),
         required=False,
         allow_none=True,
     )
 
+    base_directory = fields.String(required=False, allow_none=True)
+    glob_directive = fields.String(required=False, allow_none=True)
+    default_regex = fields.Dict(required=False, allow_none=True)
+    runtime_keys = fields.List(
+        cls_or_instance=fields.Str(), required=False, allow_none=True
+    )
+    bucket = fields.String(required=False, allow_none=True)
+    prefix = fields.String(required=False, allow_none=True)
+    delimiter = fields.String(required=False, allow_none=True)
+    max_keys = fields.Integer(required=False, allow_none=True)
+    boto3_options = fields.Dict(
+        keys=fields.Str(), values=fields.Str(), required=False, allow_none=True
+    )
+    data_asset_name_prefix = fields.String(required=False, allow_none=True)
+    data_asset_name_suffix = fields.String(required=False, allow_none=True)
+    include_schema_name = fields.Boolean(required=False, allow_none=True)
+    splitter_method = fields.String(required=False, allow_none=True)
+    splitter_kwargs = fields.Dict(required=False, allow_none=True)
+    sampling_method = fields.String(required=False, allow_none=True)
+    sampling_kwargs = fields.Dict(required=False, allow_none=True)
+    excluded_tables = fields.List(
+        cls_or_instance=fields.Str(), required=False, allow_none=True
+    )
+    included_tables = fields.List(
+        cls_or_instance=fields.Str(), required=False, allow_none=True
+    )
+    skip_inapplicable_tables = fields.Boolean(required=False, allow_none=True)
+
     @validates_schema
     def validate_schema(self, data, **kwargs):
-        pass
+        if ("default_regex" in data) and not (
+            data["class_name"]
+            in [
+                "InferredAssetFilesystemDataConnector",
+                "ConfiguredAssetFilesystemDataConnector",
+                "InferredAssetS3DataConnector",
+                "ConfiguredAssetS3DataConnector",
+            ]
+            or data["class_name"][0] == "$"
```

Comment:
What is class_name $ supposed to be?
What if someone wants to create a custom dataconnector?

Reply:
This is not a class name; this is to allow our variable substitution.  Otherwise we cannot use `class_name` as a field in `validate_schema()`.  The test of variable substitution will fail these validations when the entire config is filled with `$m`, `$g`, and similar.  So the approach I took was simply to recognize `$` as the variable substitution and allow it.

Reply:
instead of:
```
data["class_name"]
            in [
                "InferredAssetFilesystemDataConnector",
                "ConfiguredAssetFilesystemDataConnector",
                "InferredAssetS3DataConnector",
                "ConfiguredAssetS3DataConnector",
            ]
```

Let's do:

```
class_ = load_class(data["class_name"]
issubclass(class_, (<<list_of_classes>>))
```

to support users creating subclasses that can inherit configuration.

Reply:
Now I see how to read 226 "if the name of the class starts with dollar sign"